### ISSUE TITLE
Reword the Checkpointing article - misleading suggestions regarding checkpointing in distributed environment

### ISSUE THREAD

**Issue by mrembalski** [Lightning-AI/pytorch-lightning#17253]
### 📚 Documentation

The Checkpointing article suggests, that using `trainer.save_checkpoint` is the only distributed-safe solution. 

It is not entirely the case - Checkpoint callbacks, with regard to the strategy, are also distributed-safe solutions. For example, running `DDP` with ModelCheckpoint callback is a viable solution. 

The docs in question: 
```
Not using save_checkpoint() can lead to unexpected behavior and potential deadlock. Using other saving functions will result in all devices attempting to save the checkpoint.
```

An answer from @carmocca on Slack channel: 
```
carmocca
1. Each process is tied to a device in DDP
2. The devices all_reduce the weights as previously said
3. Each process runs the modelcheckpoint logic
4. trainer.save_checkpoint() is called (maybe)
5. The strategy-specific logic saves the checkpoint, in the case of DDP only rank 0 (master) does actually save
```

Slack conversation in question: 

Michał Rembalski
Hi there,
I recently stumbled upon a possible issue in PL. I have created an issue on GitHub to track it - if anyone could give it a “thumbs up” or something to propagate it further (I know it’s easy to forget about an issue with so many being open), it would be much appreciated.
https://github.com/Lightning-AI/lightning/issues/17242

william falcon
last i checked, lightning saves one checkpoint on the root node (that’s been true for years). it does not save checkpoints on the other nodes.  (edited) 

Michał Rembalski
Hi, first of all, it’s great to use PL and out of the tools available, I really like it.
The result may indeed be just one file on one node. I just checked that the function _save_checkpoint (which calls trainer.save_checkpoint inside) is actually called multiple times from different processes.
I stumbled upon it by accident (wanted to ensure that metadata of any saved checkpoint would also be saved in another place - that’s why I was looking at internal functions of ModelCheckpoint).

carmocca
The behaviour you are seeing is expected and correct. All processes run the modelcheckpoint logic and call trainer.save_checkpoint
But ultimately, only 1 checkpoint will be saved with the DDP strategy. Other strategies need to consolidate the checkpoint across ranks, hence this design

Michał Rembalski
But, if I were to run it with a “deepspeed” strategy, then the checkpoint that would be saved, could be partial, right? And no warning would be shown?

Adrian
For deepspeed it depends on the configuration, but yeah for zero3 deepspeed will save checkpoints in a folder, each rank saves one file, as this is the most important efficient way.

Michał Rembalski
That makes sense.
Then, If I wanted to use ddp strategy with multiple devices and store/delete checkpoint metadata:
The processes are synced before a checkpoint is created.
Should I do store/delete checkpoint metadata only on rank_zero or just “latest” one (or, if the processes are synced - it doesn’t matter, since the checkpoint is the same for all of them).
Also, how does this look for ddp_spawn? The docs say:
Since .spawn() trains the model in subprocesses, the model on the main process does not get updated.
Which process would save the checkpoint then? Would it be full/partial?

Jirka
[@Adrian] [@carmocca] lets update the issue too :slightly_smiling_face:

carmocca
[@Michał Rembalski] You don't need to worry about any of this! Just implement state_dict for a component and return you metadata
Note that only the state from rank zero will be saved, so if the state differs across ranks, the other rank's state will be lost. (The weights do match of course)

Michał Tyrolski
Hi 
[@carmocca] - how it works behind the scene? I mean, if each gpu worker computes gradient, then only master gpu receives all gradients, accumulate it and update weights, then share it to the rest of gpus along with updates weights, is that right? In that case state_dicts will be different among different gpus, right?

carmocca
As you said, the gradient updates are all reduced, meaning all devices will have the same version. https://pytorch.org/tutorials/beginner/ddp_series_theory.html
Then the state dict is saved. So they will match

Michał Tyrolski
ok, so checkpoint is being sent only from master node, right? Btw. thx for sharing

carmocca
1. Each process is tied to a device in DDP
2. The devices all_reduce the weights as previously said
3.  Each process runs the modelcheckpoint logic
4. trainer.save_checkpoint() is called (maybe)
5. the strategy-specific logic saves the checkpoint, in the case of DDP  only rank 0 (master) does actually save

carmocca
I updated the linked issue. Please continue in it

Michał Rembalski
[@carmocca] thanks so much for explaining it! It all makes sense now.
However, I still think that the docs could be updated, especially:
Not using save_checkpoint() can lead to unexpected behavior and potential deadlock. Using other saving functions will result in all devices attempting to save the checkpoint.
It seems a bit contrary to what you wrote - strategy-specific logic saves the checkpoint, in the case of DDP only rank 0 (master) does actually save.

carmocca
That is still correct
"Not using save_checkpoint" refers to you manually doing torch.save()

carmocca
If you do it through the trainer, the strategies take care of handling what's specific for distributed setup
but if you do torch.save , you are on your own

Michał Rembalski
I’d rather say it refers to trainer.save_checkpoint.
It is the line above it. Latest docs:
https://lightning.ai/docs/pytorch/latest/common/checkpointing_intermediate.html
I understand that Trainer’s Strategies do handle the checkpointing, even for distributed computation, but the docs state that for distributed computation you should only use trainer.save_checkpoint. At least, that’s how I understood it.
New

carmocca
then we clearly need to reword it!
Would you like to open an issue about that?

Michał Rembalski
Sure, I can do that.

cc @borda

**Comment by carmocca** [comment]
@mrembalski How do you suggest that the text is clarified? Feel free to open a PR updating the text

**Comment by mrembalski** [comment]
@carmocca i'll open the PR soon (probably today/tomorrow). I was mostly confused that there was no word about strategy of training taking part in saving the checkpoint - and most of the strategies (if not all) do support saving the checkpoints (even if they do call trainer.save_checkpoint underneath).
